ckanext/stcndm/logic/__init__.py

```diff
@@ -0,0 +1,702 @@
+__author__ = 'Statistics Canada'
+
+import functools
+import logging
+import re
+
+import formencode.validators
+
+import ckan.model as model
+import ckan.new_authz as new_authz
+import ckan.lib.navl.dictization_functions as df
+import ckan.plugins as p
+
+from ckan.common import _, c
+
+log = logging.getLogger(__name__)
+_validate = df.validate
+
+
+class AttributeDict(dict):
+    def __getattr__(self, name):
+        try:
+            return self[name]
+        except KeyError:
+            raise AttributeError('No such attribute %r' % name)
+
+    def __setattr__(self, name, value):
+        raise AttributeError(
+            'You cannot set attributes of this object directly'
+        )
+
+
+class ActionError(Exception):
+    def __init__(self, extra_msg=None):
+        self.extra_msg = extra_msg
+
+    def __str__(self):
+        err_msgs = (super(ActionError, self).__str__(),
+                    self.extra_msg)
+        return ' - '.join([str(err_msg) for err_msg in err_msgs if err_msg])
+
+
+class NotFound(ActionError):
```

Comment:
suggest importing these from ckanapi or ckan directly. Having multiple exception classes with the same name will lead to problems.